File: ios/Classes/Models/NaxaLibreMapOptions.swift
//
//  NaxaLibreMapOptions.swift
//  naxalibre
//
//  Created by Amit on 25/02/2025.
//

import Foundation


/// A struct that defines various configuration options for the NaxaLibreMap.
///
/// This struct allows customization of zoom levels, pitch limits, camera position,
/// rendering behavior, and debug options.
struct NaxaLibreMapOptions {
    
    /// The minimum zoom level allowed for the map. Default is 0.0.
    var minZoom: Double
    
    /// The maximum zoom level allowed for the map. Default is 25.5.
    var maxZoom: Double
    
    /// The minimum pitch (tilt) angle allowed for the map. Default is 0.0.
    var minPitch: Double
    
    /// The maximum pitch (tilt) angle allowed for the map. Default is 60.0.
    var maxPitch: Double
    
    /// The initial camera position of the map.
    var position: NaxaLibreMapCamera?
    
    /// The pixel ratio for rendering, useful for handling high-density displays.
    var pixelRatio: Double?
    
    /// Whether to use texture mode for rendering. This can improve performance on some devices. Default is false.
    var textureMode: Bool
    
    /// Whether debugging features (e.g., tile borders, FPS display) are enabled. Default is false.
    var debugActive: Bool
    
    /// Whether cross-source collisions are enabled for symbol placement. Default is true.
    var crossSourceCollisions: Bool
    
    /// Whether the render surface is placed on top of other UI elements. Default is false.
    var renderSurfaceOnTop: Bool
    
    /// Creates a `NaxaLibreMapOptions` instance with the specified properties.
    ///
    /// All parameters have default values to enforce explicit configuration.
    init() {
        self.minZoom = 0.0
        self.maxZoom = 25.5
        self.minPitch = 0.0
        self.maxPitch = 60.0
        self.position = nil
        self.pixelRatio = nil
        self.textureMode = false
        self.debugActive = false
        self.crossSourceCollisions = true
        self.renderSurfaceOnTop = false
    }
}

File: lib/src/models/naxalibre_map_options.dart
import 'camera_position.dart';

/// A class that defines various configuration options for the NaxaLibreMap.
///
/// This class allows customization of zoom levels, pitch limits, camera position,
/// rendering behavior, and debug options.
class NaxaLibreMapOptions {
  /// The minimum zoom level allowed for the map.
  /// Default is 0.0
  ///
  final double minZoom;

  /// The maximum zoom level allowed for the map.
  /// Default is 25.0
  ///
  final double maxZoom;

  /// The minimum pitch (tilt) angle allowed for the map.
  /// Default is 0.0
  ///
  final double minPitch;

  /// The maximum pitch (tilt) angle allowed for the map.
  /// Default is 60.0
  ///
  final double maxPitch;

  /// The initial camera position of the map.
  final CameraPosition position;

  /// The pixel ratio for rendering, useful for handling high-density displays.
  final double? pixelRatio;

  /// Whether to use texture mode for rendering. This can improve performance on some devices.
  /// Default is false
  ///
  final bool textureMode;

  /// Whether debugging features (e.g., tile borders, FPS display) are enabled.
  /// Default is false
  ///
  final bool debugActive;

  /// Whether cross-source collisions are enabled for symbol placement.
  /// Default is true
  ///
  final bool crossSourceCollisions;

  /// Whether the render surface is placed on top of other UI elements.
  /// Default is false
  ///
  final bool renderSurfaceOnTop;

  /// Creates a [NaxaLibreMapOptions] instance with the specified properties.
  ///
  /// All parameters have required named arguments to enforce explicit configuration.
  const NaxaLibreMapOptions({
    this.minZoom = 0.0,
    this.maxZoom = 25.5,
    this.minPitch = 0.0,
    this.maxPitch = 60.0,
    this.position = const CameraPosition(),
    this.pixelRatio = 0.0,
    this.textureMode = false,
    this.debugActive = false,
    this.crossSourceCollisions = true,
    this.renderSurfaceOnTop = false,
  });

  /// Converts the options into a map of arguments that can be passed to the platform channel.
  Map<String, dynamic> toArgs() {
    return {
      'minZoom': minZoom,
      'maxZoom': maxZoom,
      'minPitch': minPitch,
      'maxPitch': maxPitch,
      'position': position.toArgs(),
      'pixelRatio': pixelRatio,
      'textureMode': textureMode,
      'debugActive': debugActive,
      'crossSourceCollisions': crossSourceCollisions,
      'renderSurfaceOnTop': renderSurfaceOnTop,
    };
  }
}
